src/goto-symex/field_sensitivity.cpp

```diff
@@ -0,0 +1,321 @@
+  const namespacet &ns,
+  goto_symex_statet &state,
+  const ssa_exprt &ssa_expr) const
+{
+  if(ssa_expr.type().id() == ID_struct || ssa_expr.type().id() == ID_struct_tag)
+  {
+    const struct_typet &type = to_struct_type(ns.follow(ssa_expr.type()));
+    const struct_union_typet::componentst &components = type.components();
+
+    struct_exprt result(ssa_expr.type());
+    result.reserve_operands(components.size());
+
+    const exprt &struct_op = ssa_expr.get_original_expr();
+
+    for(const auto &comp : components)
+    {
+      const member_exprt member(struct_op, comp.get_name(), comp.type());
+      ssa_exprt tmp = ssa_expr;
+      bool was_l2 = !tmp.get_level_2().empty();
+      tmp.remove_level_2();
+      tmp.set_expression(member);
+      if(was_l2)
+        result.add_to_operands(state.rename(tmp, ns).get());
+      else
+        result.add_to_operands(get_fields(ns, state, tmp));
+    }
+
+    return std::move(result);
+  }
+#ifdef ENABLE_ARRAY_FIELD_SENSITIVITY
+  else if(
+    ssa_expr.type().id() == ID_array &&
+    to_array_type(ssa_expr.type()).size().id() == ID_constant)
+  {
+    const array_typet &type = to_array_type(ssa_expr.type());
+    const std::size_t array_size =
+      numeric_cast_v<std::size_t>(to_constant_expr(type.size()));
+
+    array_exprt result(type);
+    result.reserve_operands(array_size);
+
+    const exprt &array = ssa_expr.get_original_expr();
+
+    for(std::size_t i = 0; i < array_size; ++i)
+    {
+      const index_exprt index(array, from_integer(i, index_type()));
+      ssa_exprt tmp = ssa_expr;
+      tmp.set_expression(index);
+      result.copy_to_operands(get_fields(ns, state, tmp));
+    }
+
+    return std::move(result);
+  }
+#endif // ENABLE_ARRAY_FIELD_SENSITIVITY
+  else
+    return ssa_expr;
+}
+
+void field_sensitivityt::field_assignments(
+  const namespacet &ns,
+  goto_symex_statet &state,
+  const ssa_exprt &lhs,
+  symex_targett &target,
+  bool allow_pointer_unsoundness)
+{
+  exprt lhs_fs = lhs;
+  apply(ns, state, lhs_fs, false);
+
+  bool run_apply_bak = run_apply;
+  run_apply = false;
+  field_assignments_rec(
+    ns, state, lhs_fs, lhs, target, allow_pointer_unsoundness);
+  run_apply = run_apply_bak;
+}
+
+void field_sensitivityt::post_process_assignment(
+  const namespacet &ns,
+  goto_symex_statet &state,
+  const ssa_exprt &lhs,
+  symex_targett &target,
+  bool allow_pointer_unsoundness)
+{
+  if(is_divisible(lhs))
+  {
+    // Split composite symbol lhs into its components:
+    field_assignments(ns, state, lhs, target, allow_pointer_unsoundness);
+
+    // Erase the composite symbol from our working state
+    // (otherwise keeping both the composite and field representations risks
+    // the two becoming out of sync).
+    // Note we can't use code_deadt, which would actually try to kill the
+    // components, which we *do* want to keep!
+    //
+    // There is no need to update value sets, because they are field sensitive
```

Comment:
I think we should -- as it stands the first assignment will create `a.field1 = ..., a.field2 = ...` and then the second one will move those over to `a..field1` and `a..field2`. That *should* be harmless, as any future read should refer to the `..` field-sensitive versions, but it's at least a waste of memory and at worst going to make broken code (which somehow accidentally refers to the composite versions) appear is if it works

Reply:
But this just hits https://github.com/diffblue/cbmc/blob/f8dff52944742f7777ad763990eb101ede862b7c/src/pointer-analysis/value_set.cpp#L1203 - we could remove the `UNREACHABLE`, but it wouldn't have any effect either way?!

Reply:
I think you'd need to add a method for this -- `erase_symbol` or something which would special-case struct-typed ones to drop each field, similar to how `assign` special-cases struct-typed values.

Reply:
Done (by adding `state.value_set.values.erase(l1_identifier);`).

Reply:
That won't work (the value-set processes assignments to `composite` by decomposing them into assignments to `composite.field1`, `composite.field2`, ....). Suggest making a method on `value_sett` which applies the same decomposition logic is `value_sett::assign`

Reply:
So I reworked this a bit (getting rid of `post_process_assignment` altogether, but removing from the value set is actually not possible! Doing so breaks several C tests (it doesn't break any Java tests).

Reply:
Just to clarify what I experimented with: take a backup copy of `value_set` before `goto_symex_statet::assignment` adds information to `value_set` and restore the value set after having done field assignments.

Reply:
That copy had not taken into account the field assignments happening in between. That's now fixed, all happening in `goto_symex_statet::assignment`.